File: cli/template/packages/swift/Sources/{{Name}}Do/{{Name}}Error.swift
import Foundation

/// Errors that can occur when using {{Name}}Client
public enum {{Name}}Error: LocalizedError, Sendable {
    /// The client is not connected
    case notConnected

    /// Authentication failed
    case authenticationFailed(reason: String)

    /// Resource not found
    case notFound(entity: String, id: String)

    /// Connection failed
    case connectionFailed(reason: String)

    /// General RPC error
    case rpcError(code: String?, message: String)

    public var errorDescription: String? {
        switch self {
        case .notConnected:
            return "{{Name}}Client is not connected. Call connect() first."
        case .authenticationFailed(let reason):
            return "Authentication failed: \(reason)"
        case .notFound(let entity, let id):
            return "\(entity) not found: \(id)"
        case .connectionFailed(let reason):
            return "Connection failed: \(reason)"
        case .rpcError(let code, let message):
            if let code = code {
                return "RPC error [\(code)]: \(message)"
            }
            return "RPC error: \(message)"
        }
    }
}

File: cli/template/packages/fsharp/src/{{Name}}Error.fs
namespace {{Name}}.Do

/// Errors that can occur when using {{Name}}Client
type {{Name}}Error =
    /// The client is not connected
    | NotConnected
    /// Authentication failed
    | AuthenticationFailed of reason: string
    /// Resource not found
    | NotFound of entity: string * id: string
    /// Connection failed
    | ConnectionFailed of reason: string
    /// General RPC error
    | RpcError of code: string option * message: string

module {{Name}}Error =
    /// Convert error to human-readable message
    let message = function
        | NotConnected ->
            "{{Name}}Client is not connected. Call connectAsync() first."
        | AuthenticationFailed reason ->
            sprintf "Authentication failed: %s" reason
        | NotFound (entity, id) ->
            sprintf "%s not found: %s" entity id
        | ConnectionFailed reason ->
            sprintf "Connection failed: %s" reason
        | RpcError (Some code, msg) ->
            sprintf "RPC error [%s]: %s" code msg
        | RpcError (None, msg) ->
            sprintf "RPC error: %s" msg

/// Exception wrapper for {{Name}}Error
exception {{Name}}Exception of {{Name}}Error

module {{Name}}Exception =
    /// Raise a {{Name}}Error as an exception
    let raise error = raise ({{Name}}Exception error)
